engine/datahistory_manager.go

```diff
@@ -0,0 +1,905 @@
+		return errNilJob
+	}
+	if job.Nickname == "" {
+		return fmt.Errorf("upsert job %w", errNicknameUnset)
+	}
+
+	j, err := m.GetByNickname(job.Nickname, false)
+	if err != nil && !errors.Is(err, errJobNotFound) {
+		return err
+	}
+
+	if insertOnly && j != nil ||
+		(j != nil && j.Status != dataHistoryStatusActive) {
+		return fmt.Errorf("upsert job %w nickname: '%s' - status: '%s' ", errNicknameInUse, j.Nickname, j.Status)
+	}
+
+	m.m.Lock()
+	defer m.m.Unlock()
+
+	err = m.validateJob(job)
+	if err != nil {
+		return err
+	}
+	toUpdate := false
+	if !insertOnly {
+		for i := range m.jobs {
+			if !strings.EqualFold(m.jobs[i].Nickname, job.Nickname) {
+				continue
+			}
+			toUpdate = true
+			job.ID = m.jobs[i].ID
+			if job.Exchange != "" && m.jobs[i].Exchange != job.Exchange {
+				m.jobs[i].Exchange = job.Exchange
+			}
+			if job.Asset != "" && m.jobs[i].Asset != job.Asset {
+				m.jobs[i].Asset = job.Asset
+			}
+			if !job.Pair.IsEmpty() && !m.jobs[i].Pair.Equal(job.Pair) {
+				m.jobs[i].Pair = job.Pair
+			}
+			if !job.StartDate.IsZero() && !m.jobs[i].StartDate.Equal(job.StartDate) {
+				m.jobs[i].StartDate = job.StartDate
+			}
+			if !job.EndDate.IsZero() && !m.jobs[i].EndDate.Equal(job.EndDate) {
+				m.jobs[i].EndDate = job.EndDate
+			}
+			if job.Interval != 0 && m.jobs[i].Interval != job.Interval {
+				m.jobs[i].Interval = job.Interval
+			}
+			if job.RunBatchLimit != 0 && m.jobs[i].RunBatchLimit != job.RunBatchLimit {
+				m.jobs[i].RunBatchLimit = job.RunBatchLimit
+			}
+			if job.RequestSizeLimit != 0 && m.jobs[i].RequestSizeLimit != job.RequestSizeLimit {
+				m.jobs[i].RequestSizeLimit = job.RequestSizeLimit
+			}
+			if job.MaxRetryAttempts != 0 && m.jobs[i].MaxRetryAttempts != job.MaxRetryAttempts {
+				m.jobs[i].MaxRetryAttempts = job.MaxRetryAttempts
+			}
+			m.jobs[i].DataType = job.DataType
+			m.jobs[i].Status = job.Status
+			break
+		}
+	}
+	if job.ID == uuid.Nil {
+		job.ID, err = uuid.NewV4()
+		if err != nil {
+			return err
+		}
+	}
+	job.rangeHolder, err = kline.CalculateCandleDateRanges(job.StartDate, job.EndDate, job.Interval, uint32(job.RequestSizeLimit))
+	if err != nil {
+		return err
+	}
+
+	if !toUpdate {
+		m.jobs = append(m.jobs, job)
+	}
+
+	dbJob := m.convertJobToDBModel(job)
+	return m.jobDB.Upsert(dbJob)
+}
+
+func (m *DataHistoryManager) validateJob(job *DataHistoryJob) error {
+	if job == nil {
+		return errNilJob
+	}
+	if !job.Asset.IsValid() {
+		return asset.ErrNotSupported
+	}
+	if job.Pair.IsEmpty() {
+		return errCurrencyPairUnset
+	}
+	if !job.Status.Valid() {
+		return errInvalidDataHistoryStatus
+	}
+	if !job.DataType.Valid() {
+		return errInvalidDataHistoryDataType
+	}
+	exch := m.exchangeManager.GetExchangeByName(job.Exchange)
+	if exch == nil {
+		return fmt.Errorf("%s %w", job.Exchange, errExchangeNotLoaded)
+	}
+	pairs, err := exch.GetEnabledPairs(job.Asset)
+	if err != nil {
+		return err
+	}
+	if !pairs.Contains(job.Pair, false) {
+		return errCurrencyPairInvalid
```

Comment:
On start up had a dangling job & removed removed some enabled pairs:
[ERROR]|DATAHISTORY| 21/06/2021 15:58:56 |currency provided is not found in the available pairs list

Could be more verbose.